### PR TITLE
Special-case `.llvm` in mangler

### LINKED ISSUE #53912: [LLVM] Segfault / "llvm::thinLTOInternalizeModule ... Assertion `GS != DefinedGlobals.end()' failed".
This example (reduced from `rustc_codegen_llvm`) either crashes LLVM or makes it emit an assertion:
```rust
fn dummy() {}

mod llvm {
    pub(crate) struct Foo;
}
mod foo {
    pub(crate) struct Foo<T>(T);
    impl Foo<::llvm::Foo> {
        pub(crate) fn foo() {
            for _ in 0..0 {
                for _ in &[::dummy()] {
                    ::dummy();
                    ::dummy();
                    ::dummy();
                }
            }
        }
    }
    pub(crate) fn foo() {
        Foo::foo();
        Foo::foo();
    }
}
pub fn foo() {
    foo::foo();
}
```

I suspect the `for` loops aren't needed, but the control-flow there affects the bug in finicky ways.

The `llvm` name of the module appears to be important and tied to the fact that we end up emitting `llvm::` in a symbol name (because of the `impl`), which gets mangled as `llvm..` (note the two dots).
Among other things, changing the `.` we use to mangle `:`, to `$` makes the bug go away.

Assertion message (if LLVM assertions are enabled):
```cxx
rustc: /checkout/src/llvm/lib/Transforms/IPO/FunctionImport.cpp:961:
auto llvm::thinLTOInternalizeModule(llvm::Module &, const llvm::GVSummaryMapTy &)::(anonymous class)::operator()(const llvm::GlobalValue &) const:
Assertion `GS != DefinedGlobals.end()' failed.
```

Stack backtrace for SIGSEGV (if LLVM assertions are disabled):
```cxx
#0  0x00007fffed5240d2 in std::_Function_handler<bool (llvm::GlobalValue const&), llvm::thinLTOInternalizeModule(llvm::Module&, llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*> > const&)::$_2>::_M_invoke(std::_Any_data const&, llvm::GlobalValue const&) ()
   from /home/eddy/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so
#1  0x00007fffed5409be in llvm::InternalizePass::maybeInternalize(llvm::GlobalValue&, std::set<llvm::Comdat const*, std::less<llvm::Comdat const*>, std::allocator<llvm::Comdat const*> > const&) ()
   from /home/eddy/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so
#2  0x00007fffed540e26 in llvm::InternalizePass::internalizeModule(llvm::Module&, llvm::CallGraph*) ()
   from /home/eddy/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so
#3  0x00007fffed520079 in llvm::thinLTOInternalizeModule(llvm::Module&, llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*> > const&) ()
   from /home/eddy/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so
#4  0x00007fffecda34f8 in LLVMRustPrepareThinLTOInternalize ()
   from /home/eddy/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so
#5  0x00007fffeccb1a2c in rustc_codegen_llvm::back::lto::LtoModuleCodegen::optimize ()
   from /home/eddy/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so
#6  0x00007fffecc5fbb2 in rustc_codegen_llvm::back::write::execute_work_item ()
   from /home/eddy/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends/librustc_codegen_llvm-llvm.so
```

cc @sunfishcode @denismerigoux @alexcrichton @rkruppe

### LINKED ISSUE #60925: [LLVM] Segmentation fault on MacOS with > 1 codegen units and optimization > 0
I'm seeing a segmentation fault in the [Weld project](https://github.com/weld-project/weld/tree/segfault) during compilation when compiling with `cargo build --release`. This is only happening on MacOS (tested on High Sierra and Mojave); it works fine on Ubuntu 16.04. This issue also does not occur without `--release`.

I haven't been able to make a smaller example for this, but it seems to have started after we renamed the `llvm2` submodule to `llvm` (https://github.com/weld-project/weld/pull/440). 

After some investigation, we found that _any one  of_ the following changes fix the segmentation fault, but are non-ideal:

* Add `opt-level = 0` to `profile.release` in the Cargo manifest.
* Add `codegen-units = 1` to `profile.release` in the Cargo manifest.
* Rename the `weld::codegen::llvm` to something else (e.g., `weld::codegen::llvm2`): this is what seems to have introduced the issue.

This is almost certainly related to #53912, but the code example in that issue seems fixed: I'll update this issue with a minimal example when I can come up with one.

The faulting command is:

```
rustc --edition=2018 --crate-name weld weld/src/lib.rs --color always --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=597bae5b2ab31e15 -C extra-filename=-597bae5b2ab31e15 --out-dir /Users/shoumikpalkar/work/weld/target/release/deps -L dependency=/Users/shoumikpalkar/work/weld/target/release/deps --extern chrono=/Users/shoumikpalkar/work/weld/target/release/deps/libchrono-f261edcedff7fdf9.rlib --extern env_logger=/Users/shoumikpalkar/work/weld/target/release/deps/libenv_logger-48ad719d69c29b83.rlib --extern fnv=/Users/shoumikpalkar/work/weld/target/release/deps/libfnv-c5377afc6353fb2e.rlib --extern lazy_static=/Users/shoumikpalkar/work/weld/target/release/deps/liblazy_static-7ef694354b396823.rlib --extern libc=/Users/shoumikpalkar/work/weld/target/release/deps/liblibc-841557f1e9c82484.rlib --extern llvm_sys=/Users/shoumikpalkar/work/weld/target/release/deps/libllvm_sys-b77778a82d9c4c52.rlib --extern log=/Users/shoumikpalkar/work/weld/target/release/deps/liblog-ac2f0859500275e8.rlib --extern num_integer=/Users/shoumikpalkar/work/weld/target/release/deps/libnum_integer-9a9190e54f393ad7.rlib --extern regex=/Users/shoumikpalkar/work/weld/target/release/deps/libregex-c35e629b59596f2c.rlib --extern time=/Users/shoumikpalkar/work/weld/target/release/deps/libtime-6604aa2275081d44.rlib --extern uuid=/Users/shoumikpalkar/work/weld/target/release/deps/libuuid-7f660aa24d3f6e4a.rlib -C link-args=-Wl,-export_dynamic -L native=/Users/shoumikpalkar/work/weld/target/release/build/weld-c91c7a8f561133ca/out -l z -l c++ -l dylib=stdc++ -l static=llvmext -L native=/Users/shoumikpalkar/work/weld/target/release/build/llvm-sys-681bcfd36d11b4f1/out -L 'native=/usr/local/Cellar/llvm@6/6.0.1_1/lib'```
```

And LLDB gives the following information:

```
* thread #4, stop reason = EXC_BAD_ACCESS (code=1, address=0xffffffffffffff41)
    frame #0: 0x0000000106a7575d librustc_codegen_llvm-llvm.dylib`std::__1::__function::__func<llvm::thinLTOInternalizeModule(llvm::Module&, llvm::DenseMap<unsigned long long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long long>, llvm::detail::DenseMapPair<unsigned long long, llvm::GlobalValueSummary*> > const&)::$_2, std::__1::allocator<llvm::thinLTOInternalizeModule(llvm::Module&, llvm::DenseMap<unsigned long long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long long>, llvm::detail::DenseMapPair<unsigned long long, llvm::GlobalValueSummary*> > const&)::$_2>, bool (llvm::GlobalValue const&)>::operator()(llvm::GlobalValue const&) + 765
librustc_codegen_llvm-llvm.dylib`std::__1::__function::__func<llvm::thinLTOInternalizeModule(llvm::Module&, llvm::DenseMap<unsigned long long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long long>, llvm::detail::DenseMapPair<unsigned long long, llvm::GlobalValueSummary*> > const&)::$_2, std::__1::allocator<llvm::thinLTOInternalizeModule(llvm::Module&, llvm::DenseMap<unsigned long long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long long>, llvm::detail::DenseMapPair<unsigned long long, llvm::GlobalValueSummary*> > const&)::$_2>, bool (llvm::GlobalValue const&)>::operator():
->  0x106a7575d <+765>: movzbl 0xc(%rax), %eax
    0x106a75761 <+769>: andl   $0xf, %eax
    0x106a75764 <+772>: addl   $-0x7, %eax
    0x106a75767 <+775>: cmpl   $0x1, %eax
Target 0: (rustc) stopped.
```

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #60925 and fixes #53912.

r? @michaelwoerister 
cc @eddyb

### CODE REVIEW

**Comment by davidtwco** [comment]
I believe this is what @eddyb suggested. I wasn't able to work out how to write a test for this, so any thoughts appreciated.

**Comment by michaelwoerister** [comment]
Thanks for looking into it, @davidtwco! That should fix the issue.

@bors r+

**Comment by michaelwoerister** [comment]
Good call, bors... Regarding a test: You need to write some code with a module or function named `llvm`, I'd say, and then test that the symbol name does not contain the string `.llvm`. [ui/symbol-names](https://github.com/rust-lang/rust/tree/master/src/test/ui/symbol-names] show one way of checking symbol names. A `codegen` would be another option.

**Comment by eddyb** [comment]
https://github.com/rust-lang/rust/issues/53912 has a reduction (I guess it's tricky to cause it because the segfault is in LTO).

**Comment by davidtwco** [comment]
Thanks @eddyb and @michaelwoerister. I've added two tests, one that checks the symbol name with the reproduction from #53912 and another that checks that the code compiles successfully (since I wasn't sure that the symbol check attribute didn't exit early).

**Review by eddyb — APPROVED** [review]

File: src/librustc_codegen_utils/symbol_names.rs [hunk 1]
```diff
@@ -613,6 +613,9 @@ impl fmt::Write for SymbolPrinter<'_, '_> {
                 // for ':' and '-'
                 '-' | ':' => self.path.temp_buf.push('.'),
 
+                // Avoid segmentation fault on some platforms, see #60925.
+                'm' if self.path.temp_buf.ends_with(".llv") => self.path.temp_buf.push_str("$6d$"),
```
Comment (eddyb) [inline src/librustc_codegen_utils/symbol_names.rs]: Ooops, turns out this should be `$u6d$` (i.e. note the `u`).

Comment (davidtwco) [reply src/librustc_codegen_utils/symbol_names.rs]: I've submitted #61423 that changes this.

**Pull request merged**